airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java

```diff
@@ -1,169 +1,236 @@
-            FileUtil.closeQuietly(songWriter);
-        }
-    }
-
-}
+/*
+ This file is part of Airsonic.
+
+ Airsonic is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Airsonic is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Airsonic.  If not, see <http://www.gnu.org/licenses/>.
+
+ Copyright 2016 (C) Airsonic Authors
+ Based upon Subsonic, Copyright 2009 (C) Sindre Mehus
+ */
+
+package org.airsonic.player.service.search;
+
+import org.airsonic.player.domain.Album;
+import org.airsonic.player.domain.Artist;
+import org.airsonic.player.domain.MediaFile;
+import org.airsonic.player.domain.MusicFolder;
+import org.airsonic.player.util.FileUtil;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.SearcherManager;
+import org.apache.lucene.store.FSDirectory;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+/**
+ * Function class that is strongly linked to the lucene index implementation.
+ * Legacy has an implementation in SearchService.
+ * 
+ * If the index CRUD and search functionality are in the same class,
+ * there is often a dependency conflict on the class used.
+ * Although the interface of SearchService is left to maintain the legacy implementation,
+ * it is desirable that methods of index operations other than search essentially use this class directly.
+ */
+@Component
+public class IndexManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(IndexManager.class);
+
+    // XXX 3.x -> 8.x : #1166 Holds a tentative literal. It needs to be resolved by the release.
+    private static final String INDEX_DIR = "lucene2";
+
+    @Autowired
+    private AnalyzerFactory analyzerFactory;
+
+    @Autowired
+    private DocumentFactory documentFactory;
+
+    @Autowired
+    private SearchServiceUtilities util;
+
+    private Map<IndexType, SearcherManager> searchers = new ConcurrentHashMap<>();
+
+    private Map<IndexType, IndexWriter> writers = new ConcurrentHashMap<>();
```

Comment:
Can you explain the need for the concurrent hash map here? Can you detail which entrypoints/methods into IndexManager may be called concurrently?

Reply:
I only keep in mind add/remove for map.
(Such as multiple calls from REST)
Lucene classes such as SearcherManager are thread safe.